opentelemetry-prometheus/src/lib.rs

```diff
@@ -197,7 +199,65 @@ struct CollectorInner {
     metric_families: HashMap<String, MetricFamily>,
 }
 
+// TODO: Remove lazy and switch to pattern matching once `TypeId` is stable in
+// const context: https://github.com/rust-lang/rust/issues/77125
+static HISTOGRAM_TYPES: Lazy<[TypeId; 3]> = Lazy::new(|| {
+    [
+        TypeId::of::<data::Histogram<i64>>(),
+        TypeId::of::<data::Histogram<u64>>(),
+        TypeId::of::<data::Histogram<f64>>(),
+    ]
+});
+static SUM_TYPES: Lazy<[TypeId; 3]> = Lazy::new(|| {
+    [
+        TypeId::of::<data::Sum<i64>>(),
+        TypeId::of::<data::Sum<u64>>(),
+        TypeId::of::<data::Sum<f64>>(),
+    ]
+});
+static GAUGE_TYPES: Lazy<[TypeId; 3]> = Lazy::new(|| {
+    [
+        TypeId::of::<data::Gauge<i64>>(),
+        TypeId::of::<data::Gauge<u64>>(),
+        TypeId::of::<data::Gauge<f64>>(),
+    ]
+});
+
 impl Collector {
+    fn metric_type_and_name(&self, m: &data::Metric) -> Option<(MetricType, Cow<'static, str>)> {
```

Comment:
Can we move this later when `add_*` functions? It may save us one downcast/runtime type check. I do see this version is easier to understand and less repetitive though but wonder if the performance wise if it make sense to avoid repeating downcast when the type is `Sum`

Reply:
We could, but I don't think the downcasting has any noticeable perf impact. Could do a benchmark but threads like https://users.rust-lang.org/t/downcasting-overhead/69718/6 make it pretty clear.

Reply:
Doesn't seem to make a difference when I run quick bench, can re-assess if perf becomes an issue later.

Reply:
Sg. Thanks for checking!